<a href="https://colab.research.google.com/github/FengHu101/FengHu101/blob/main/Feng_Hu_shared_competition2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Conclusion：I used the decision tree classification algorithm to get a score of 0.2425, which I think is the closest to the ideal score of all the classifier algorithms I have used.

Code referred from https://colab.research.google.com/drive/1pJOtTqNsRplBbc-Sk8DVvVYyNbFkOV4o?usp=sharing#scrollTo=Kn3g00zlnjkC

In [ ]:
import pandas as pd #initializing
df_train = pd.read_csv('SBSP-train.csv') #loading dataset to python dataframe
df_test = pd.read_csv('SBSP-test.csv')
#df_test['label'] = 0
df_train.head()

label                                          utterance
0      2  Give me a break. You don't know anything about...
1      2  Ugh, nail-biting is a disgusting bad habit, so...
2      0                                    But, Mr. Krabs.
3      3  Yeah, me neither. No matter how cool and shiny...
4      0                      I can do this! I can do this!

In [ ]:
X_train = df_train['utterance'] #set x,y for training
y_train = df_train['label']
X_test = df_test['utterance'] #set x,y for testing
y_test = df_test['label'] 

The spliting of two sets of original data is to select the model with the best effect (which can be understood as accuracy rate) and the best generalization ability.

In [ ]:
from sklearn.model_selection import train_test_split 
X_train, X_test, y_train, y_test = train_test_split(X_train, y_train, test_size=0.2) #spliting dataset to X_train, y_train, and 20% for testing

In the process of machine learning, we often need to vectorize data, that is, generate some specific vector and then train and calculate.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer  #Vectorzing data....
vectorizer = CountVectorizer(max_features=1000)
X_train_counts = vectorizer.fit_transform(X_train).toarray() #transform data to matrix
X_test_counts = vectorizer.transform(X_test).toarray()

Code referred from https://colab.research.google.com/drive/1ygxseB4veI_jcMv5PLEBCXkE838_EJhD?usp=sharing#scrollTo=s5rpwTX6QjUZ

In [ ]:
from sklearn.naive_bayes import MultinomialNB #Using MultinmialNB classifier to training the data
model = MultinomialNB()
model.fit(X_train_counts, y_train) #fit data...

MultinomialNB()

In [ ]:
model.score(X_test_counts, y_test) #get the accurate valuse from training

0.4663888888888889

The training score of the MultinomialNB algorithm is about 0.407, which is far from the ideal score, indicating that this algorithm is not applicable to this data set.

Let's try to use Perceptron for training... Perceptron is a dichotomous linear model whose inputs are feature vectors and outputs are categories. The role of Perceptron is to divide data into hyperplanes of positive and negative types. It's the most basic classifier in machine learning.
https://colab.research.google.com/drive/1nDpIpnZDsBmqlAXr19OSe3UHXDA2WY_m#scrollTo=zEOxPh9nByXo

In [ ]:
from sklearn.linear_model import Perceptron #Using Perceptron to training data
model = Perceptron()
model.fit(X_train_counts, y_train)
model.score(X_test_counts, y_test)

0.4077777777777778

The perceptron algorithm's score is about 0.4078. It is more ideal than very high grade MultinomialNB. However, it is not the most suitable classification algorithm for this dataset.

Let's train the data set with the following MLPClassifier, setting the number of iterations of the data to 50.
https://colab.research.google.com/drive/1pJOtTqNsRplBbc-Sk8DVvVYyNbFkOV4o?usp=sharing#scrollTo=Kn3g00zlnjkC

In [ ]:
from sklearn.neural_network import MLPClassifier  #Using MLPClassifier to training data
model = MLPClassifier(max_iter=50)
model.fit(X_train_counts, y_train)


/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (50) reached and the optimization hasn't converged yet.
  warnings.warn(


MLPClassifier(max_iter=50)

In [ ]:
model.score(X_test_counts, y_test)

0.4608333333333333

The score of this training is not very ideal too, and it does not seem to be the most suitable classifier

In [ ]:
vectorizer2 = CountVectorizer(max_features=1000,ngram_range=(1,2)) #Make second vector of the data set
X_train_counts = vectorizer2.fit_transform(X_train).toarray()
X_test_counts = vectorizer2.transform(X_test).toarray()

In machine learning, the function of vectorizing data multiple times is to improve the efficiency of computation. The function of vectorization of data for many times is to further improve the efficiency of calculation by constantly splitting calculation tasks and using multiple computing units for calculation. In this way, we can greatly reduce the calculation time and improve the training speed of the model while ensuring the accuracy of the calculation. https://colab.research.google.com/drive/1ygxseB4veI_jcMv5PLEBCXkE838_EJhD?usp=sharing#scrollTo=z6HUmfE_AjGP

In [ ]:
model = Perceptron()
model.fit(X_train_counts, y_train)

Perceptron()

In [ ]:
model.score(X_test_counts, y_test)

0.43833333333333335

After the second vectorization, the result is still not very ideal, so we have to continue to find a more suitable classifier.

In [ ]:
from sklearn.tree import DecisionTreeClassifier #Using DecisionTreeClassifier to training data
clf = DecisionTreeClassifier(criterion='entropy',max_depth=0.5)
clf.fit(X_train_counts, y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=0.5)

Decision tree algorithm is one of the most classic algorithms in machine learning, it is easy to understand and implement: decision tree model adopts tree structure, model interpretation is good, algorithm implementation is simple. And can deal with data missing value, you can not need to preprocess the data. https://colab.research.google.com/drive/1nDpIpnZDsBmqlAXr19OSe3UHXDA2WY_m#scrollTo=zEOxPh9nByXo

In [ ]:
clf.score(X_test_counts, y_test)

0.2425

I think I got the perfect score, which was close to 0.25.

In [ ]:
from sklearn.neighbors import KNeighborsClassifier #Using KNN Classifier to training data
knn = KNeighborsClassifier(n_neighbors=5,p=2)
knn.fit(X_train_counts, y_train)

KNeighborsClassifier()

KNN classifier is also a very classical machine learning algorithm. I also used it to train and test the data set, but it did not get an ideal score. Stochastic gradient descent classification is also a very suitable algorithm for large data sets. It only needs to load a small batch of data to quickly train the model, but I did not get an ideal score from it。
https://colab.research.google.com/drive/16iSxLAxF-826Jp2lPEZKLrUXQWqu2mR5?usp=sharing

In [ ]:
knn.score(X_test_counts, y_test)

0.3377777777777778

In [ ]:
from sklearn.linear_model import SGDClassifier #Using SDG Classifier to training data
sgd = SGDClassifier(max_iter=2000, tol = 1e-3, random_state=42)
sgd.fit(X_train_counts, y_train)

SGDClassifier(max_iter=2000, random_state=42)

In [ ]:
sgd.score(X_test_counts, y_test)

0.4577777777777778